File: src/main/java/cube8540/oauth/authentication/users/application/DefaultUserService.java
package cube8540.oauth.authentication.users.application;

import cube8540.oauth.authentication.users.domain.User;
import cube8540.oauth.authentication.users.domain.UserAuthority;
import cube8540.oauth.authentication.users.domain.UserRepository;
import cube8540.oauth.authentication.users.domain.Username;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Service;

import java.util.Collections;
import java.util.Set;
import java.util.stream.Collectors;

@Service
public class DefaultUserService implements UserDetailsService {

    private final UserRepository repository;

    @Autowired
    public DefaultUserService(UserRepository repository) {
        this.repository = repository;
    }

    @Override
    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
        User user = repository.findByUsername(new Username(username))
                .orElseThrow(() -> new UsernameNotFoundException(username + " is not found"));

        Set<GrantedAuthority> authorities = user.getAuthorities().stream()
                .map(UserAuthority::getValue).map(SimpleGrantedAuthority::new)
                .collect(Collectors.toSet());

        return org.springframework.security.core.userdetails.User.builder()
                .username(user.getUsername().getValue()).password(user.getPassword())
                .accountLocked(!user.isCredentials()).authorities(Collections.emptySet())
                .authorities(authorities)
                .build();
    }
}

File: src/main/kotlin/cube8540/oauth/authentication/users/application/DefaultUserService.kt
package cube8540.oauth.authentication.users.application

import cube8540.oauth.authentication.users.domain.UserRepository
import cube8540.oauth.authentication.users.domain.Username
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.core.userdetails.User
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Service
class DefaultUserService @Autowired constructor(
    private val repository: UserRepository
): UserDetailsService {

    @Transactional(readOnly = true)
    override fun loadUserByUsername(username: String): UserDetails {
        val user = repository.findById(Username(username))
            .orElseThrow { UsernameNotFoundException("$username is not found") }

        return User.builder()
            .username(user.username.value).password(user.password)
            .accountLocked(!user.credentialed)
            .authorities(Collections.emptyList())
            .build()
    }
}